### PR TITLE
Implement pre-projection

### LINKED ISSUE #1500: Projection pre and post order
Right now we have issues with projection removing items but still having decorators run on them or reference those items.

## Proposal

1. Define 2 stage for projection to run `pre` and `post`
2. Allow to define projection for member elements

Currently the flow for a projection is:

1. Start projecting type
2. Recursively project children
3. Run decorators
4. Run the projection (Here we might remove the type and now decorators have already been run on this type and its children)

With the proposed changed we insert a new step:

1. Start projecting type
2. Run `pre` projection (Here we would decide to remove the type and so we don't run the decorator nor have the children created.)
3. Recursively project children
4. Run decorators
5. Run `post` projection

## Syntax change

```cadl
projection model#v {
  to {
    // This remains the `post` order projection
  }

  pre to {
    // Here define the step that run pre.
  }

  // Alternative name
  preTo {

  }
}
```

## Target member elements

Use reflection types?

```cadl
projection Cadl.Reflection.ModelProperty#v {

}
```

### LINKED ISSUE #1708: Add a property to Interface metadata to indicate extending another interface
Implementation issue for: https://github.com/microsoft/typespec/issues/1636

- Add a property to interface metadata to rindicate extending another interface
- Property should be similar to properties in model metadata for extending models
- Documentation should make it clear that the property does not imply the idea of inheritance for the interface

### PR DESCRIPTION
**Pull request by nguerrera**
- Add support for `pre to` and `pre from` projections that can run before children are projected and return never to delete or a different type. 

- It is an error to mutate anything pre-projections but this is not currently enforced by the compiler. We had a note to document this, but projections are still experimental and undocumented, so documeting it here.

- Add support for `#enummember`, `#unionvariant`, and `#modelproperty` projections.

- Use these features in versioning library to fix issues that led to emitter crashes with versioned resource types.

- Fix issues with incompatible versioniing verification:
  - Bogus error asking to annotate a template definition when passing a versioned template argument.
  - Failure to validate versioned types in tuples or unions.
  - Failure to validate template arguments/array element type in operation return type.

- Add access to the extends list of an interface via `Interface.sourceInterfaces`. This was a requested feature for next sprint, but ended up helping to get the validation fixed so it is done here.

Fix #1500
Fix #1708

### CODE REVIEW

File: packages/compiler/core/parser.ts [hunk 1]
```diff
@@ -1547,50 +1550,64 @@ function createParser(code: string | SourceFile, options: ParseOptions = {}): Pa
...
-    if (token() === Token.Identifier) {
-      proj1 = parseProjection();
 
-      if (token() === Token.Identifier) {
-        proj2 = parseProjection();
+    const projectionMap = new Map<string, ProjectionNode>();
+    const projections: ProjectionNode[] = [];
+    while (token() === Token.Identifier) {
+      const projection = parseProjection();
+      if (projection.direction !== "<error>") {
```
Comment (timotheeguerin) [inline packages/compiler/core/parser.ts:1558]: Does `direction` still make sense as the name?

File: packages/compiler/core/types.ts [hunk 2]
```diff
@@ -281,6 +281,16 @@ export interface Interface extends BaseType, DecoratedType, TemplatedTypeBase {
...
 
+  /**
+   * The interfaces that provide additional operations via `interface extends`.
+   *
+   * Note that despite the same `extends` keyword in source form, this is a
+   * different semantic relationship than the one from {@link Model} to
+   * {@link Model.baseModel}. Operations from extended interfaces are copied
+   * into {@link Interface.operations}.
+   */
+  sourceInterfaces: Interface[];
```
Comment (timotheeguerin) [inline packages/compiler/core/types.ts:292]: Is this related to this change?

File: packages/compiler/core/projector.ts [hunk 3]
```diff
@@ -589,19 +594,49 @@ export function createProjector(
...
+      const targetNode =
+        projectionApplication.direction === "from"
+          ? projectionsByName[0].preFrom
+          : projectionsByName[0].preTo;
+
+      if (targetNode) {
+        const projected = checker.project(type, targetNode, projectionApplication.arguments);
+        if (projected !== type) {
+          // override the projected type cache with the returned type
+          projectedTypes.set(type, projected);
```
Comment (timotheeguerin) [inline packages/compiler/core/projector.ts:631]:
```suggestion
          projectedTypes.set(type, projected);
```

would this not cause any issues with referencing the previous type if you already made a reference to it?

File: packages/compiler/core/parser.ts (hunk 1 above)
Comment (nguerrera) [reply packages/compiler/core/parser.ts:1558]: It's a bit off. I wanted to avoid renaming property and breaking and also couldn't come up with something better.

File: packages/compiler/core/types.ts (hunk 2 above)
Comment (nguerrera) [reply packages/compiler/core/types.ts:292]: Yes, it's also a distinct feature that was on the schedule for next sprint, but I realized I needed it to get the validation working. Once I correctly fixed validation to not fire on template instantiations themselves, I needed a way to still fire on the reference to the instantiation in `interface Foo extends Bar<NewerThanFoo>`. We had similar handling for model instantiations, but nowhere clean to get this reference to interface instantiation so I added the feature now and used it.

File: packages/compiler/core/parser.ts (hunk 1 above)
Comment (timotheeguerin) [reply packages/compiler/core/parser.ts:1558]: don't think thats a big deal so whatever but would also having a `pre: boolean` property instead help?

File: packages/compiler/core/projector.ts (hunk 3 above)
Comment (nguerrera) [reply packages/compiler/core/projector.ts:631]: Yeah, you're right, this seems wrong. I wrote this so long ago and this got put aside until we resolved the design questions that I forget what the intent was here. I will look and add more tests.

Comment (nguerrera) [reply packages/compiler/core/projector.ts:631]: I don't think there was any intent, it looks like I left it mistakenly by copying from applyProjection to applyPreProjection.

Comment (nguerrera) [reply packages/compiler/core/projector.ts:631]: I have opened a can of worms testing this. Will file an issue. There are pre-existing problems, I think with return `<non-intrinsic>`.

**Review by timotheeguerin — APPROVED** [review]

File: packages/compiler/core/parser.ts (hunk 1 above)
Comment (nguerrera) [reply packages/compiler/core/parser.ts:1571]: Thought about this but it ends up being helpful to have a single differentiating "key" so leaving it.
